Use multiple field/value terms in ES query

BartChris opened the pull request:
This PR addresses point 8) of https://github.com/kitodo/kitodo-production/issues/6743. The problem with the current implementation is, that when you have a search like` projectname:MDP_LIK `to retrieve all processes which have a metadata value of "MDP_LIK" in the project field, Kitodo first tokenizes the term to "MDP" and "LIK" and then issues **two** requests for Elasticsearch. 
Both requests return independent sets of IDs. And when both sets do not intersect we might get less hits then we actually have.

For each token constructed at query time we inject the list of IDs (retrieved from the Index) into SQL, so at the end we have sth like `WHERE id IN (x,y,z) and ID in (a,b,c)`. The individual ID lists might be huge because they contain all results which just contain the fragment "MDP".

My change does ensure that only one query is sent to the search index and only one list of IDs is returned, so we do not get multiple contradicting ID lists. It also in many cases ensures that the ID lists are way shorter because we already intersect at the Elasticsearch level and not just pass huge ID lists to the database which are not relevant.

This also frees us from having to construct complex SQL restrictions for multiple returned ID list. Since we only have one list we can just append a simple `WHERE id in (<list od ids>)`.

BartChris commented on Kitodo/src/main/java/org/kitodo/production/services/index/IndexingService.java:
```diff
@@ -123,24 +125,43 @@
+        logger.debug(
+                "Searching {} IDs with terms {}: {} hits",
+                beanClass.getSimpleName(),
+                termSummary,
```
It seems like we are not allowed to log the terms somebody searched for in the debug log. I will show only the search Keys for now.

BartChris commented on Kitodo/src/main/java/org/kitodo/production/services/index/IndexingService.java:
```diff
@@ -126,25 +127,45 @@ public CompletionStage<?> startIndexing(Class<? extends BaseBean> type, MassInde
+            var query = searchSession.search(beanClass)
+                    .select(idField)
+                    .where(f -> {
+                        var bool = f.bool();
```
You have a point, i will try to clarify. This structure is idiomatic Hibernate Search DSL:
https://docs.hibernate.org/search/6.2/reference/en-US/html_single/#query-predicate

matthias-ronge reviewed (commented):
There is only one question for me to clarify, if you can say that this isn’t a problem, then I approve the PR.

matthias-ronge commented on Kitodo/src/main/java/org/kitodo/production/services/data/IndexQueryPart.java:
```diff
@@ -88,33 +84,14 @@ public FilterField getFilterField() {
-            for (String lookingFor : lookfor) {
-                queryCount++;
-                String uniqueParameterName = parameterName + UNIQUE_PARAMETER_EXTENSION + queryCount;
-                restrictions.add(varName + "." + idField + (operand ? " IN (:" : " NOT IN (:") + uniqueParameterName + ')');
```
Handling of NOT search (`operand`) was removed and has no replacement.

matthias-ronge commented on Kitodo/src/main/java/org/kitodo/production/services/data/ProcessService.java:
```diff
@@ -297,7 +297,7 @@ private BeanQuery createProcessQuery(Map<?, String> filters, boolean showClosedP
             project -> showInactiveProjects || project.isActive()).map(Project::getId).collect(Collectors.toList());
         query.restrictToProjects(projectIDs);
-        query.performIndexSearches();
+        query.applyIndexRestriction("id");
```
Please use a static final string for all occurrences of `"id"`.

matthias-ronge commented on Kitodo/src/main/java/org/kitodo/production/services/data/TaskService.java:
```diff
@@ -177,7 +177,7 @@ public Long countResults(Map<?, String> filters, boolean onlyOwnTasks, boolean h
 
         BeanQuery query = formBeanQuery(filters, onlyOwnTasks, hideCorrectionTasks, showAutomaticTasks, taskStatus);
-        query.performIndexSearches();
+        query.applyIndexRestriction("process.id");
```
Please use a static final string for both occurrences of `"process.id"`.

matthias-ronge reviewed (changes requested):
The NOT search (value of `operand`) is lost.

solth commented:
@BartChris please rebase against `main` to resolve conflicts!

BartChris committed:
Improve index request handling

BartChris committed:
Implement requested changes

solth commented on Kitodo/src/main/java/org/kitodo/production/services/data/ProcessService.java:
```diff
@@ -142,6 +142,7 @@
 public class ProcessService extends BaseBeanService<Process, ProcessDAO> {
 
     private static final Map<String, String> SORT_FIELD_MAPPING;
+    private static final String FIELD_ID = "id";
```
In the future such generic constants - whose values are used in many other classes as well - could be moved to `Kitodo-API/src/main/java/org/kitodo/constants/StringConstants.java`

BartChris commented:
> The NOT search (value of `operand`) is lost.

Very good point, will have to look deeper here. A negation index query is right now not documented. https://github.com/kitodo/kitodo-production/wiki/Suchen-und-filtern

BartChris committed:
Support NOT in index queries

BartChris commented:
I added back the option to prefix metadata searches with "-", e.g. `-TitleDocMain:Urkunde` to allow for "Does not contain"-searches. To pass the information (field, value, operand) around i introduced a Java `record`.
My hope is that we can reduce a lot of the complexity invoved in passing information between the DB search and Index search when switching to a DB only search.

matthias-ronge commented:
> A negation index query is right now not documented. https://github.com/kitodo/kitodo-production/wiki/Suchen-und-filtern

It is only mentioned here: https://github.com/kitodo/kitodo-production/wiki/Suche-und-filtern#ausschlie%C3%9Fende-suche

matthias-ronge commented on Kitodo/src/main/java/org/kitodo/production/services/data/IndexQueryTerm.java:
```diff
@@ -0,0 +1,18 @@
+
+package org.kitodo.production.services.data;
+
+public record IndexQueryTerm(
```
Please add javadoc.

```suggestion
/**
 * A token of an index search.
 * 
 * @param field
 *            search field selected by the user
 * @param token
 *            index query term
 * @param operand
 *            whether the search must match
 */
public record IndexQueryTerm(
```

matthias-ronge reviewed (approved):
This is OK now. You may want to add Javadoc to the `record`.

solth commented:
@BartChris would you mind rebasing your branch against `main` one more time?
